Add support for Guava RangeMap class [DRAFT]

mcvayc opened the pull request:
This commit fixes https://github.com/FasterXML/jackson-datatypes-collections/issues/195 by supporting [Guava RangeMap](https://guava.dev/releases/20.0/api/docs/com/google/common/collect/RangeMap.html) serialization and deserialization for `RangeMaps` that use a `Range` that supports String serialization.  Suggestions and comments welcome.

mcvayc committed:
Add support for Guava RangeMap class

mcvayc commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/GuavaKeyDeserializers.java:
```diff
@@ -0,0 +1,22 @@
+    static final long serialVersionUID = 1L;
+
+    @Override
+    public KeyDeserializer findKeyDeserializer(JavaType type, DeserializationConfig config, BeanDescription beanDesc) throws JsonMappingException {
```
RangeMaps use `Range`s as the key for a `Map`-like object so we need to define a KeyDeserializer for `Range` objects.

mcvayc commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/deser/RangeKeyDeserializer.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public KeyDeserializer createContextual(DeserializationContext ctxt,
```
Nearly all the code in this class is "borrowed" from `guava/src/main/java/com/fasterxml/jackson/datatype/guava/deser/RangeDeserializer.java`.  This is definitely not as DRY as I want it to be but I wasn't sure what the standard is for Jackson projects for code sharing/reuse.  Is it preferred to create a util class with public static methods?  Or should I create a base class that has the code that's shared between `RangeDeserializer` and `RangeKeyDeserializer`?  What's the recommended approach?  Let me know and I'll make the change.

cowtowncoder replied:
Alas, I don't think there is a really good way: both `KeyDeserializer` and `JsonDeserializer` (in 3.0 `ValueDeserializer`) are abstract classes so implementations can at most extend shared interface (but no base class).
For utility class may be the way to go.

cowtowncoder replied:
Ok, I can see the issue -- lots of copied code.

Yeah, I would try creating a util/helper class first; helper class could be non-static, if that helps (constructed with config of value/key deserializer)

mcvayc replied:
Thanks for the guidance.  I'll take that approach.

mcvayc commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/deser/RangeMapDeserializer.java:
```diff
@@ -0,0 +1,261 @@
+     */
+    private final Method creatorMethod;
+
+    public RangeMapDeserializer(MapLikeType type, KeyDeserializer keyDeserializer,
```
I used the `Multimap` deserializer as guide for the `RangeMap` deserializer.  It's probably possible to clean-up somethings in here.

mcvayc commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/ser/RangeMapSerializer.java:
```diff
@@ -0,0 +1,444 @@
+ * Serializer for Guava's {@link RangeMap} values. Output format encloses all
+ * values in JSON Map.
+ */
+public class RangeMapSerializer
```
Similarly, I used the Multimap serializer as guide for the RangeMap serializer so more clean-up might be possible here.

cowtowncoder commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/deser/RangeKeyDeserializer.java:
```diff
@@ -0,0 +1,132 @@
+ * use Delegating Deserializer, using a POJO as an intermediate form (properties
+ * could be of type {@link Object})
+ * This would also also simplify the implementation a bit.
+ */
```
`@since 2.21` here and elsewhere for new classes

cowtowncoder commented on guava/src/main/java/com/fasterxml/jackson/datatype/guava/deser/RangeKeyDeserializer.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    public Object deserializeKey(String rangeInterval, DeserializationContext context) throws IOException {
+        if (rangeInterval.isEmpty()) {
```
Do we want to allow use of empty Strings as `null`s? As opposed to indicating an error.
(if we do, need to make sure to test in unit tests).

cowtowncoder committed:
Merge branch '2.x' into 2.x

cowtowncoder commented:
@mcvayc Looks very good so far! 

Now... I hate to ask this but would it be possible at all to split this into 2 PRs -- starting with serialization, to be merged first, and then deser to follow?
The reason I am asking is because merging from 2.x to 3.x will be quite a bit of work, due to Jackson internal API changes: I can handle that, but it would be easier to do in 2 parts; merge to 2.x working piece, I merge forward to 3.x, fix issues, then 2.x deser change, same.

mcvayc commented:
> @mcvayc Looks very good so far!
> 
> Now... I hate to ask this but would it be possible at all to split this into 2 PRs -- starting with serialization, to be merged first, and then deser to follow? The reason I am asking is because merging from 2.x to 3.x will be quite a bit of work, due to Jackson internal API changes: I can handle that, but it would be easier to do in 2 parts; merge to 2.x working piece, I merge forward to 3.x, fix issues, then 2.x deser change, same.

No problem.  I'll split this up.

mcvayc committed:
Address PR comments.

- Refactor code to reduce duplication.
- Add test for filters
- Throw exception on empty Range key values.

mcvayc commented:
@cowtowncoder I've moved the shared code into `RangeHelper.java`, added an exception for empty string keys (since guava doesn't support them for RangeMap), updated/added the version tag, and added a test for filtering.

If this looks good to you, I will close this PR and open two new ones in its place.  I figured it would be easier to review the whole thing together first though, since I had to do some refactoring.

cowtowncoder commented:
Looks good @mcvayc ! I must admit I did not read all parts in detail since they are copied/modeled after existing implementations, but did not see any obvious issues.

I also think we already have CLA as well.

So, split to serialize first, then deserialize (some tests may need to wait for second part) would be great.
